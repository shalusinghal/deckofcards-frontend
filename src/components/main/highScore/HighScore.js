import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

const HighScore = (props) => {
  const { highScoreList } = props;
  let renderList = (
    <tr>
      <th colSpan="3" className="text-center">No data available</th>
    </tr>
  );

  if (highScoreList.length > 0) {
    renderList = highScoreList.map((score, index) => (
      <tr key={`${score.turn}_${score.name}`}>
        <th scope="row">{index + 1}</th>
        <td>{score.turn}</td>
        <td>{score.name}</td>
      </tr>
    ));
  }
  return (
    <div className="row mt-3">
      <div className="col-md-12">
        <h3>High Score List</h3>
        <table className="table table-striped">
          <thead>
            <tr>
              <th scope="col">#</th>
              <th scope="col">Turn</th>
              <th scope="col">Name</th>
            </tr>
          </thead>
          <tbody>
            {renderList}
          </tbody>
        </table>
      </div>
    </div>
  );
};

HighScore.propTypes = {
  highScoreList: PropTypes.oneOfType([PropTypes.array]).isRequired,
};

const mapStateToProps = (state) => ({
  highScoreList: state.highScore.highScoreList,
});

const mapDispatchToProps = {};

const HighScoreContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(HighScore);

export default HighScoreContainer;
