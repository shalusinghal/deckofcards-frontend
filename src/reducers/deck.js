import { ACTION_TYPES, CARD } from '../constants';

const prepareDecks = (action, state) => {
  const { deckIds, totalItems } = action;

  const card = {
    frontSrc: CARD.frontSrc,
    backSrc: CARD.backSrc,
    displayState: 'front',
    alreadyDrawn: false,
  };

  const newDeck = deckIds.map((deckId) => {
    let cards = [];
    // Add the card and code for each deck
    for (let index = 0; index < totalItems; index += 1) {
      cards = [...cards, { ...card, code: `${index}_${deckId}` }];
    }

    return { deckId, cards };
  });

  return {
    ...state,
    decks: newDeck,
    cardDrawCounter: 0,
    match: 0,
    lastOpenCard: null,
  };
};

const changeDisplayStateOfCard = (deck, index) => (
  deck.cards.map((card, counter) => {
    if (counter === index) {
      return { ...card, displayState: (card.displayState === 'front') ? 'back' : 'front' };
    }
    return card;
  })
);

const doToggleCard = (action, state) => {
  const { deckId, index } = action;
  const decks = [...state.decks];
  const newDecks = decks.map((deck) => {
    if (deck.deckId === deckId) {
      // Change the display state
      const cards = changeDisplayStateOfCard(deck, index);
      return { deckId, cards };
    }
    return deck;
  });

  return { ...state, decks: newDecks };
};

const setNewCard = (action, state) => {
  const { deckId, newCard, index } = action.options;
  const decks = [...state.decks];

  const newDecks = decks.map((deck) => {
    if (deck.deckId === deckId) {
      // set the new card to same index and remaiing will be same
      const newDeck = { ...deck };
      const { cards } = newDeck;
      cards[index] = newCard;

      return { deckId, cards: newDeck.cards };
    }
    return deck;
  });

  return { ...state, decks: newDecks };
};

const setLastOpenCard = (action, state) => {
  const { deckId, index } = action.options;

  // return null if last open card is already there
  if (state.lastOpenCard !== null) {
    return { ...state, lastOpenCard: null };
  }

  // Get the selected deck and card and set it as last open card
  const selectedDeck = state.decks.find((deck) => deck.deckId === deckId);
  const newLastOpenCard = selectedDeck.cards.find((card, cardIndex) => cardIndex === index);
  newLastOpenCard.deckId = deckId;
  newLastOpenCard.index = index;

  return { ...state, lastOpenCard: newLastOpenCard };
};

const initialState = {
  decks: [],
  cardDrawCounter: 0,
  match: 0,
  lastOpenCard: null,
};

const deck = (state = initialState, action) => {
  switch (action.type) {
    case ACTION_TYPES.shuffle:
      return prepareDecks(action, state);
    case ACTION_TYPES.toggleCard:
      return doToggleCard(action, state);
    case ACTION_TYPES.newDraw:
      return setNewCard(action, state);
    case ACTION_TYPES.lastOpenCard:
      return setLastOpenCard(action, state);
    case ACTION_TYPES.cardDrawCounter:
      return { ...state, cardDrawCounter: state.cardDrawCounter + 1 };
    case ACTION_TYPES.cardMatch:
      return { ...state, match: state.match + 1 };
    default:
      return state;
  }
};

export default deck;
