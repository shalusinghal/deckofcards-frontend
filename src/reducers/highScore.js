import { ACTION_TYPES } from '../constants';

const highScore = (state = { highScoreList: [] }, action) => {
  if (action.type === ACTION_TYPES.fetchHighScore) {
    return { highScoreList: action.highScoreList };
  }

  return state;
};

export default highScore;
