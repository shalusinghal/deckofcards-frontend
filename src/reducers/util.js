import { ACTION_TYPES } from '../constants';

const util = (state = { loading: false, isError: false }, action) => {
  switch (action.type) {
    case ACTION_TYPES.loading:
      return { ...state, loading: !state.loading };
    case ACTION_TYPES.error:
      return { ...state, isError: action.errorFlag };
    default:
      return state;
  }
};

export default util;
