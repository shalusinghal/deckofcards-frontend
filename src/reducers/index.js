import { combineReducers } from 'redux';

import highScore from './highScore';
import util from './util';
import deck from './deck';

const gameApp = combineReducers({
  deck,
  highScore,
  util,
});

export default gameApp;
