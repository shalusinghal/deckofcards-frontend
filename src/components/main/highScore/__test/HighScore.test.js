import React from 'react';
import { createStore } from 'redux';
import renderer from 'react-test-renderer';

import reducers from '../../../../reducers';
import HighScore from '../HighScore';

const store = createStore(reducers);

it('renders correctly when there are no data', () => {
  const tree = renderer.create(<HighScore store={store} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly when data is available', () => {
  const highScoreList = [{ name: 'Shalu SInghal', turn: 12 }];
  const tree = renderer.create(<HighScore store={store} highScoreList={highScoreList} />).toJSON();
  expect(tree).toMatchSnapshot();
});
