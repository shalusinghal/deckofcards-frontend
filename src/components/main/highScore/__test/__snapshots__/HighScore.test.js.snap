// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when data is available 1`] = `
<div
  className="row mt-3"
>
  <div
    className="col-md-12"
  >
    <h3>
      High Score List
    </h3>
    <table
      className="table table-striped"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            #
          </th>
          <th
            scope="col"
          >
            Turn
          </th>
          <th
            scope="col"
          >
            Name
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            className="text-center"
            colSpan="3"
          >
            No data available
          </th>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders correctly when there are no data 1`] = `
<div
  className="row mt-3"
>
  <div
    className="col-md-12"
  >
    <h3>
      High Score List
    </h3>
    <table
      className="table table-striped"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            #
          </th>
          <th
            scope="col"
          >
            Turn
          </th>
          <th
            scope="col"
          >
            Name
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th
            className="text-center"
            colSpan="3"
          >
            No data available
          </th>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
